src/main/java/me/steinborn/krypton/mixin/shared/network/flushconsolidation/ClientConnectionMixin.java

```diff
@@ -88,12 +88,22 @@ public Channel disableForcedFlushEveryTick(ClientConnection clientConnection) {
         return null;
     }
 
-    private void doSendPacket(Packet<?> packet, @Nullable GenericFutureListener<? extends Future<? super Void>> callback) {
+    private void doSendPacket(Packet<?> packet, class_7648 callback) {
         if (callback == null) {
             this.channel.write(packet, this.channel.voidPromise());
         } else {
             ChannelFuture channelFuture = this.channel.write(packet);
-            channelFuture.addListener(callback);
+            channelFuture.addListener(listener -> {
+              if (listener.isSuccess()) {
+                callback.method_45083(); // onSuccess moj
+              } else {
+                Packet<?> failedPacket = callback.method_45086(); // onFailure moj
+                if (failedPacket != null) {
+                  ChannelFuture failedChannelFuture = this.channel.writeAndFlush(failedPacket);
```

Comment:
Should skip flush on write and check `this.autoFlush` to flush it?

```suggestion
                  ChannelFuture failedChannelFuture = this.channel.write(failedPacket, this.channel.voidPromise());
```

Reply:
I don't think forcing the flush here will cause much harm.